lint/ktlint.bzl

```diff
@@ -0,0 +1,157 @@
+"""API for calling declaring an ktlint lint aspect.
+
+Typical usage:
+Make sure you have `ktlint` pulled as a dependency into your WORKSPACE/module by pulling a version of it from here
+https://github.com/pinterest/ktlint/releases and using a `http_file` declaration for it like.
+
+```
+http_file(
+    name = "com_github_pinterest_ktlint",
+    sha256 = "2e28cf46c27d38076bf63beeba0bdef6a845688d6c5dccd26505ce876094eb92",
+    url = "https://github.com/pinterest/ktlint/releases/download/1.2.1/ktlint",
+    executable = True,
+)
+```
+
+Then, create the linter aspect, typically in `tools/lint/linters.bzl`:
+
+```starlark
+load("@aspect_rules_lint//lint:ktlint.bzl", "ktlint_aspect")
+
+ktlint = ktlint_aspect(
+    binary = "@@com_github_pinterest_ktlint//file",
+    # rules can be enabled/disabled from with this file
+    editorconfig = "@@//:.editorconfig",
+    # a baseline file with exceptions for violations
+    baseline_file = "@@//:.ktlint-baseline.xml",
+)
+```
+"""
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")
+load("//lint/private:lint_aspect.bzl", "LintOptionsInfo", "filter_srcs", "report_file")
+
+_MNEMONIC = "ktlint"
+
+def ktlint_action(ctx, executable, srcs, editorconfig, report, baseline_file, java_runtime, use_exit_code = False):
+    """ Runs ktlint as build action in Bazel.
+
+    Adapter for wrapping Bazel around
+    https://pinterest.github.io/ktlint/latest/install/cli/
+
+    Args:
+        ctx: an action context or aspect context
+        executable: struct with ktlint field
+        srcs: A list of source files to lint
+        editorconfig: The file object pointing to the editorconfig file used by ktlint
+        report: :output:  the stdout of ktlint containing any violations found
+        baseline_file: The file object pointing to the baseline file used by ktlint.
+        java_runtime: The Java Runtime configured for this build, pulled from the registered toolchain.
+        use_exit_code: whether a non-zero exit code from ktlint process will result in a build failure.
+    """
+
+    args = ctx.actions.args()
+    inputs = srcs
+    outputs = [report]
+
+    # ktlint artifact is published as an "executable" script which calls the fat jar
+    # so we need to pass a hermetic Java runtime from our build to avoid relying on
+    # system Java
+    java_home = java_runtime[java_common.JavaRuntimeInfo].java_home
```

Comment:
Hm, this isn't working on my machine:

```
alex@a:~/Projects/rules_lint/example$ bazel build '--toolchain_resolution_debug=.*java.*' src:hello_kt --output_groups=rules_lint_report --aspects=//tools/lint:linters.bzl%ktlint
INFO: ToolchainResolution: Performing resolution of @@bazel_tools//tools/jdk:runtime_toolchain_type for target platform @@local_config_platform//:host
      ToolchainResolution:   Toolchain @@rules_java~7.1.0~toolchains~local_jdk//:jdk is compatible with target plaform, searching for execution platforms:
      ToolchainResolution:     Compatible execution platform @@local_config_platform//:host
      ToolchainResolution:   All execution platforms have been assigned a @@bazel_tools//tools/jdk:runtime_toolchain_type toolchain, stopping
      ToolchainResolution: Recap of selected @@bazel_tools//tools/jdk:runtime_toolchain_type toolchains for target platform @@local_config_platform//:host:
      ToolchainResolution:   Selected @@rules_java~7.1.0~toolchains~local_jdk//:jdk to run on execution platform @@local_config_platform//:host
INFO: ToolchainResolution: Target platform @@local_config_platform//:host: Selected execution platform @@local_config_platform//:host, type @@bazel_tools//tools/jdk:runtime_toolchain_type -> toolchain @@rules_java~7.1.0~toolchains~local_jdk//:jdk
INFO: Analyzed target //src:hello_kt (0 packages loaded, 7 targets configured).
INFO: From ktlint src/ktlint.hello_kt.aspect_rules_lint.report:
external/com_github_pinterest_ktlint/file/downloaded: 10: [: Illegal number: 
external/com_github_pinterest_ktlint/file/downloaded: 12: exec: java: not found
ERROR: /home/alex/Projects/rules_lint/example/src/BUILD.bazel:57:15: output 'src/ktlint.hello_kt.aspect_rules_lint.report' was not created
ERROR: /home/alex/Projects/rules_lint/example/src/BUILD.bazel:57:15: ktlint src/ktlint.hello_kt.aspect_rules_lint.report failed: not all outputs were created or valid
Aspect //tools/lint:linters.bzl%ktlint of //src:hello_kt failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.071s, Critical Path: 0.01s
INFO: 2 processes: 1 internal, 1 linux-sandbox.
ERROR: Build did NOT complete successfully
```

Reply:
That distributable self-extracting executable starts with

```
#!/bin/sh

# From this SO answer: https://stackoverflow.com/a/56243046

# First we get the major Java version as an integer, e.g. 8, 11, 16. It has special handling for the leading 1 of older java
# versions, e.g. 1.8 = Java 8
JV=$(java -version 2>&1 | sed -E -n 's/.* version "([^.-]*).*".*/\1/p')

# Add --add-opens for java version 16 and above
X=$( [ "$JV" -ge "16" ] && echo "--add-opens java.base/java.lang=ALL-UNNAMED" || echo "")

exec java $X -Xmx512m -jar "$0" "$@"
```

so I think we need `java` on the `$PATH` as well. Simply setting `PATH` in the env doesn't work as then the script doesn't find `sed`.

Reply:
Ah, interesting. I guess it works on my machine because I have `java` in `PATH`. Makes sense to add the `java` from the toolchain to `PATH` as well.

Reply:
Reproduced failure on CI here as well: https://github.com/aspect-build/rules_lint/actions/runs/8802281517/job/24157603293?pr=215#step:5:214

Reply:
I can reproduce it locally now. Will try to push a commit to fix.

Reply:
Interestingly it works with `--java_language_version=17` and `--java_runtime_version=17`, and not with JDK 21 which seems to be the default registered toolchain.

Reply:
Actually my bad, I forgot local JDK is the default toolchain and my Java 21 installation was broken. It works after installing it correctly now and also works with `remotejdk_21`.

Reply:
I get an error locally on linux:

```
INFO: From ktlint src/ktlint.hello_kt.aspect_rules_lint.report:
Exception in thread "main" java.lang.NoClassDefFoundError: javax/swing/Icon
        at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
        at java.base/java.lang.Class.privateGetDeclaredConstructors(Unknown Source)
        at java.base/java.lang.Class.getConstructor0(Unknown Source)
        at java.base/java.lang.Class.getConstructor(Unknown Source)
        at org.jetbrains.kotlin.psi.stubs.elements.KtStubElementType.<init>(KtStubElementType.java:43)
        at org.jetbrains.kotlin.psi.stubs.elements.KtClassElementType.<init>(KtClassElementType.java:42)
        at org.jetbrains.kotlin.psi.stubs.elements.KtStubElementTypes.<clinit>(KtStubElementTypes.java:32)
        at org.jetbrains.kotlin.KtNodeTypes.<clinit>(KtNodeTypes.java:28)
        at com.pinterest.ktlint.rule.engine.core.api.ElementType.<clinit>(ElementType.kt:16)
        at com.pinterest.ktlint.ruleset.standard.rules.AnnotationRule.<clinit>(AnnotationRule.kt:476)
        at com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider$getRuleProviders$1.invoke(StandardRuleSetProvider.kt:107)
        at com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider$getRuleProviders$1.invoke(StandardRuleSetProvider.kt:107)
        at com.pinterest.ktlint.rule.engine.core.api.RuleProvider$Companion.invoke(RuleProvider.kt:41)
        at com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider.getRuleProviders(StandardRuleSetProvider.kt:107)
        at com.pinterest.ktlint.cli.internal.LoadRuleProvidersKt.loadRuleProviders(LoadRuleProviders.kt:18)
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.getRuleProviders$ktlint_cli(KtlintCommandLine.kt:370)
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.lintOrFormat(KtlintCommandLine.kt:269)
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.run(KtlintCommandLine.kt:244)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:279)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:41)
        at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:457)
        at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:454)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:474)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:481)
        at com.pinterest.ktlint.Main.main(Main.kt:20)
Caused by: java.lang.ClassNotFoundException: javax.swing.Icon
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
        ... 25 more
16:27:06.436 [main] INFO com.pinterest.ktlint.cli.internal.KtlintCommandLine -- Enable default patterns [**/*.kt, **/*.kts]
```
but as long as CI is green I'm willing to accept that it's my own problem.

Reply:
I think there's one integration test failing
```
not ok 2 should fail when --fail-on-violation is passed
# (from function `assert_output' in file /usr/lib/bats/bats-assert/src/assert_output.bash, line 186,
#  from function `assert_lints' in file test/lint_test.bats, line 26,
#  in test file test/lint_test.bats, line [7](https://github.com/aspect-build/rules_lint/actions/runs/8808592582/job/24177926484?pr=215#step:5:8)4)
#   `assert_lints' failed
#
# -- output does not contain substring --
# substring (1 lines):
#   src/hello.kt:1:1: File name 'hello.kt' should conform PascalCase (standard:filename)
# output (6[9](https://github.com/aspect-build/rules_lint/actions/runs/8808592582/job/24177926484?pr=215#step:5:10) lines):
```
I guess when violations are enforced, the output should be directed to std out as well as report (right now it's just the latter). Let me fix it.

Reply:
Right, you could use this pattern verbatim
https://github.com/aspect-build/rules_lint/blob/main/lint/shellcheck.bzl#L43-L46

Reply:
Thanks!!